lib/livepush.ts

```diff
@@ -0,0 +1,112 @@
+import * as Bluebird from 'bluebird';
+import * as Dockerode from 'dockerode';
+import { EventEmitter } from 'events';
+import * as _ from 'lodash';
+import StrictEventEmitter from 'strict-event-emitter-types';
+
+import Container, { CommandOutput, StageContainers } from './container';
+import Dockerfile from './dockerfile';
+import { InvalidArgumentError } from './errors';
+
+export interface LivepushEvents {
+	commandExecute: { stageIdx: number; command: string };
+	commandOutput: { stageIdx: number; output: CommandOutput };
+	commandReturn: { stageIdx: number; returnCode: number };
+	containerRestart: { containerId: string };
+}
+
+type ContainerEventEmitter = StrictEventEmitter<EventEmitter, LivepushEvents>;
+
+export class Livepush extends (EventEmitter as {
+	// We need to avoid the tslint errors here, as typescript
+	// will not accept the changes proposed
+	// tslint:disable-next-line
+	new (): ContainerEventEmitter;
+}) {
+	private constructor(
+		public docker: Dockerode,
+		public dockerfile: Dockerfile,
+		public containers: StageContainers,
+	) {
+		super();
+		this.assignEventHandlers();
+	}
+
+	public static async init(
+		dockerfileContent: string | Buffer,
+		context: string,
+		containerId: string,
+		stageImages: string[],
+		docker: Dockerode,
+	): Promise<Livepush> {
+		const dockerfile = new Dockerfile(dockerfileContent);
+
+		if (dockerfile.stages.length - 1 !== stageImages.length) {
+			const dStages = dockerfile.stages.length;
+			const argStages = stageImages.length;
+			throw new InvalidArgumentError(
+				`Dockerfile with ${dStages} stages provided,` +
+					` but ${argStages} image IDs passed to livepush constructor (there should be ${dStages -
+						1})`,
+			);
+		}
+
+		const containers: StageContainers = {};
+		// create the list of containers, in the order of the
+		// stages
+		for (const [idx, stageImage] of stageImages.entries()) {
+			containers[idx] = await Container.fromImage(context, docker, stageImage);
+		}
+		containers[dockerfile.stages.length - 1] = Container.fromContainerId(
+			context,
+			docker,
+			containerId,
+		);
+
+		return new Livepush(docker, dockerfile, containers);
+	}
+
+	public async performLivepush(
+		addedOrUpdated: string[],
+		deleted: string[],
+	): Promise<void> {
+		const tasks = this.dockerfile.getActionGroupsFromChangedFiles(
+			addedOrUpdated.concat(deleted),
+		);
+
+		const keys = _.keys(tasks).sort();
+		await Bluebird.each(keys, async (stageIdxStr: string) => {
+			const stageIdx = parseInt(stageIdxStr, 10);
+			const stageTasks = tasks[stageIdx];
+
+			await this.containers[stageIdx].executeActionGroups(
```

Comment:
Does this need to be `return await ...`?

Reply:
This is now no longer necessary due to the switching of the `Bluebird.each` to a `for..of`